Core: Improve startup events

tmeasday opened the pull request:
Issue: #10893 

## What I did (so far)
 - [x] Add a new `StoreSelectionSpecifier` concept (id, name+kind, or '*')
 - [x] Set specifier right away based on URL in preview
 - [x] At end of configuring (before `SET_STORIES`), make a selection based on specifier (which emits `CURRENT_STORY_WAS_SET`).
 - [x] Change render to listen for `CURRENT_STORY_WAS_SET` and act appropriately.
 - [x] No longer select first story in manager on startup if no story is chosen in URL.
 - [x] Listen to `CURRENT_STORY_WAS_SET` and select story in manager
 - [x] Handle URLS for "first story of component" and a storyId that is a non-leaf.
 - [x] Ensure the first *sorted* story is selected
 - [x] Update tests

tmeasday committed:
Update preview to work in new scheme

tmeasday committed:
Don't select first story in manager any more

tmeasday commented:
@ndelangen - some questions before I take this further:

1. This implementation emits `CURRENT_STORY_WAS_SET` just *before* `SET_STORIES`. It could equally do it after but this seems right to me. Do you anticipate any problems?

2. Do you think we need to support the two (I think esoteric?) features: a) `path=/story/componentId--*`, b) `path=/story/groupId` -- where are these used, do you recall?

3. How does `navigate()` handle being called with the existing URL? Is that OK? (What's going to happen when the user makes a selection in the UI is: a) manager calls `SELECT_STORY`, b) manager calls `navigate(newStoryId)`, c) manager emits `SET_CURRENT_STORY`, d) store changes, emits `CURRENT_STORY_WAS_SET`, e) manager calls `navigate(newStoryId)`. Does that seem OK?

ndelangen commented:
> 1. This implementation emits `CURRENT_STORY_WAS_SET` just *before* `SET_STORIES`. It could equally do it after but this seems right to me. Do you anticipate any problems?

Yeah I think that makes more sense
SET_STORIES >> CURRENT_STORY_WAS_SET

> 2. Do you think we need to support the two (I think esoteric?) features: a) `path=/story/componentId--*`, b) `path=/story/groupId` -- where are these used, do you recall? Supporting them is probably OK, I just want to make sure it is necessary first.

I think we don't need this

> 3. How does `navigate()` handle being called a second time with the current URL/storyId? Is that OK? (What's going to happen when the user makes a selection in the UI is: a) manager calls `SELECT_STORY`, b) manager calls `navigate(newStoryId)`, c) manager emits `SET_CURRENT_STORY`, d) store changes, emits `CURRENT_STORY_WAS_SET`, e) manager calls `navigate(newStoryId)`. Does that seem OK? -- if not we'll enter a loop). If it's not OK is there a way to check if the navigation is necessary or not?

The preview will ignore SET_CURRENT_STORY if it's already on that story.

We could optimize a bit more and not emit SET_CURRENT_STORY in that case maybe.
I don't think we'll end up in a infinite loop, because of the reason stated above.

ndelangen commented:
does CURRENT_STORY_WAS_SET have data on the event? We should add it if not.

CURRENT_STORY_WAS_SET could be an aggregation of STORY_RENDERED STORY_ERRORED STORY_MISSING ?

tmeasday commented:
> Yeah I think that makes more sense
> SET_STORIES >> CURRENT_STORY_WAS_SET

Sorry, what does the second line here mean?

>> Do you think we need to support the two (I think esoteric?) features: a) path=/story/componentId--*, b) path=/story/groupId -- where are these used, do you recall? Supporting them is probably OK, I just want to make sure it is necessary first.
> I think we don't need this

OK I'm going to drop it then.

> does CURRENT_STORY_WAS_SET have data on the event? We should add it if not.

It has the selection. It is the corresponding event to `SET_CURRENT_STORY` and has the same data. It's the event that's emitted in response to `SET_CURRENT_STORY` after the selection was made.

> CURRENT_STORY_WAS_SET could be an aggregation of STORY_RENDERED STORY_ERRORED STORY_MISSING ?

No, because it happens *before* rendering (it is what triggers the `StoryRenderer` to render). It's probably OK though because the `StoryRenderer` will render syncronously when the event is emitted in the preview.

shilman commented:
> Do you think we need to support the two (I think esoteric?) features: a) path=/story/componentId--*, b) path=/story/groupId -- where are these used, do you recall? Supporting them is probably OK, I just want to make sure it is necessary first.

This was a commonly-requested feature back when we implemented it. Users wanted to be able to link to components without having to worry about which story they are linking to. I think we first implemented (a) and then later some user thought it was ugly and added (b). I think we should drop (a) and support (b) if it's not difficult.

tmeasday committed:
Update StoryStore with full selection specifier API

tmeasday committed:
Fix test

tmeasday committed:
Navigate in manager on CURRENT_STORY_WAS_SET

tmeasday commented on lib/api/src/tests/stories.test.js:
```diff
@@ -405,95 +410,33 @@ describe('stories API', () => {
-        api: { setStories },
-      } = initStories({ store, navigate, provider });
+  //     init();
+  //     api.emit(CURRENT_STORY_WAS_SET, { storyId: 'a--1', viewMode: 'story' });
```
@ndelangen -- how can I emit an event here that is treated as coming from the local ref? Do you have a pattern for this in tests?

ndelangen replied:
The `this` ref is added here:
https://github.com/storybookjs/storybook/blob/8a20d1c178cd7aeb1acb41bd07accfde9ddb54bc/lib/channel-postmessage/src/index.ts#L43-L44

Testing this would require a mod on the EventEmitter..

We could do this:
https://github.com/primus/EventEmitter3#contextual-emits

tmeasday replied:
I found a decent solution

ndelangen replied:
LGTM

tmeasday reviewed (commented):
@ndelangen this works and is ready for review. There is just two more tests I am trying to get done.

tmeasday committed:
Fix StoryRenderer test

tmeasday committed:
Update stories tests to call CURRENT_STORY_WAS_SET

tmeasday committed:
Linting

tmeasday committed:
Fix up url tests

ndelangen commented on lib/api/src/modules/stories.ts:
```diff
@@ -294,6 +255,20 @@ export const init: ModuleFn = ({
+    // Later when we change story via the manager (or SELECT_STORY below), we'll already be at the
+    // correct path before CURRENT_STORY_WAS_SET is emitted, so this is less important (the navigate is a no-op)
+    // Note this is the case for refs also.
+    fullAPI.on(CURRENT_STORY_WAS_SET, function handleCurrentStoryWasSet({ storyId, viewMode }) {
+      const { source }: { source: string } = this;
+      const [sourceType] = getSourceType(source);
+
+      if (sourceType === 'local') {
+        navigate(`/${viewMode}/${storyId}`);
+      }
+    });
```
Looks good!

ndelangen commented on lib/core-events/src/index.ts:
```diff
@@ -6,10 +6,6 @@ enum events {
   CURRENT_STORY_WAS_SET = 'currentStoryWasSet',
   // Emitted by the preview whenever the list of stories changes (in batches)
   SET_STORIES = 'setStories',
-  // Emitted by the preview when the current story (or error) should be rendered
-  RENDER_CURRENT_STORY = 'renderCurrentStory',
-  // Deprecated version of the above
-  STORY_RENDER = 'storyRender',
```
This is a major breaking change for addons, we should document in migration.md?

tmeasday replied:
OK!

tmeasday replied:
Done: https://github.com/storybookjs/storybook/pull/11080/commits/ea86077196a0b8ae20ba00169a1ac0623213e394

ndelangen commented on lib/core/src/client/preview/url.test.ts:
```diff
@@ -52,39 +46,6 @@ describe('url', () => {
     });
   });
 
-  describe('getIdFromLegacyQuery', () => {
-    const getRawStory = () => {};
-    const store: unknown = { getRawStory };
-    it('should parse story paths', () => {
-      expect(getIdFromLegacyQuery({ path: '/story/story--id' }, store as StoryStore)).toBe(
-        'story--id'
-      );
-    });
-    it('should use legacy parameters to look up custom story ids', () => {
-      const customStore: unknown = {
-        getRawStory: () => ({ id: 'custom--id' }),
-      };
-      expect(
-        getIdFromLegacyQuery(
-          { selectedKind: 'kind', selectedStory: 'story' },
-          customStore as StoryStore
-        )
-      ).toBe('custom--id');
-    });
-    it('should use fall-back behavior for legacy queries for unknown stories', () => {
-      expect(
-        getIdFromLegacyQuery(
-          { path: null, selectedKind: 'kind', selectedStory: 'story' },
-          store as StoryStore
-        )
-      ).toBe('kind--story');
-    });
-
-    it('should not parse non-queries', () => {
-      expect(getIdFromLegacyQuery({}, store as StoryStore)).toBeUndefined();
-    });
-  });
```
So glad to see this code go!

tmeasday replied:
Well actually it isn't really gone: https://github.com/storybookjs/storybook/pull/11111

ndelangen commented:
👏👏👏👏👏👏👏 @tmeasday

tmeasday committed:
Add deprecation message for selectedName/Kind urls

tmeasday committed:
Check window parent isn't window (i.e not really a parent)

tmeasday committed:
Update documentation about startup evetns

tmeasday committed:
Merge pull request #11111 from storybookjs/10960-deprecate-legacy-url-scheme

Add deprecation message for selectedName/Kind urls

tmeasday committed:
Merge pull request #11112 from storybookjs/check-window-global

Check window parent isn't window (i.e not really a parent)

ndelangen commented:
I'm available if you need me

tmeasday committed:
Linting

tmeasday committed:
Fix story renderer tests

tmeasday committed:
Fix up test that got messed up somehow

tmeasday committed:
Delay screenshots for addon-queryparams

tmeasday committed:
Linting

tmeasday committed:
Fix storyshots issue

tmeasday committed:
Not sure why I would need this

ndelangen committed:
FIX channel-postmessage check for window === window.parent

ndelangen committed:
FIX versions

ndelangen committed:
Merge branch 'next' into 10893-refactor-startup-events

tmeasday commented:
@ndelangen can you remind me how to reproduce these e2e test failures?

tmeasday committed:
Fix issue with addon-queryparams

It was replacing `location.search` but that could be the empty string :facepalm:

tmeasday committed:
Remove pointless delay

tmeasday committed:
Merge remote-tracking branch 'origin/next' into 10893-refactor-startup-events

ndelangen committed:
Merge branch 'next' into 10893-refactor-startup-events

tmeasday commented:
@ndelangen?

shilman commented on lib/client-api/src/story_store.test.ts:
```diff
@@ -468,21 +482,174 @@ describe('preview.story_store', () => {
+
+        expect(store.getSelection()).toEqual({ storyId: 'a--1', viewMode: 'story' });
+      });
+      it('takes into account sorting', () => {
```
```suggestion

      it('takes into account sorting', () => {
```

shilman committed:
Update lib/client-api/src/story_store.test.ts

shilman reviewed (approved):
GREAT!
